### PR TITLE
Classes for adding kube config and kube cluster picker to Controller connection dialog

### PR DESCRIPTION
**Pull request by ranasaria**
**Description:**
This change adds classes and objects needed for adding the following the controller connection dialog:
1. kube config file picker
2. cluster context picker from within the kube config
The PR also includes unit tests for the classes and functions.
All of the product code is copied from resource-deployment with some refactoring and pruning. The tests have been added a new. In a future work we will expose objects from the arc extensions using extension api from the arc extension in resource-deployment and then we will remove the duplicate code from there.

This PR is first in a series of PRs to address: #12727

**Testing:**
Extension unit tests.

### CODE REVIEW

File: extensions/arc/src/test/ui/components/radioOptionsGroup.test.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	});
+
+	afterEach(() => {
+		sinon.restore();
+	});
+
+	it('verify construction and load', async () => {
+		should(radioOptionsGroup).not.be.undefined();
+		should(radioOptionsGroup.value).not.be.undefined();
+		radioOptionsGroup.value!.should.equal('value2');
```
Comment (Charles-Gagnon) [inline extensions/arc/src/test/ui/components/radioOptionsGroup.test.ts:36]: Add message explaining that the default value should be set to value2

File: extensions/arc/src/test/ui/components/radioOptionsGroup.test.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		// verify all the radioButtons created in the group
+		verifyRadioGroup('value2');
+	});
+
+	it('onClick', async () => {
+		//click the radioButton corresponding to 'value1'
+		radioButtons.filter(r => r.value === 'value1').pop()!.click();
```
Comment (Charles-Gagnon) [inline extensions/arc/src/test/ui/components/radioOptionsGroup.test.ts:43]: Yeah so tying the specific test implementation of the object creation together like this isn't great. You should be having generic functions that create the objects but then leave it up to the logic you're testing to actually hook stuff up and just have the specific tests do the checks to make sure that things were added as expected.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+};
+
+export const loadingComponent = <azdata.LoadingComponent>{
+	get loading() { return false; },
+	set loading(_value: boolean) { }
+};
+
+const loadingBuilder = (<azdata.LoadingComponentBuilder><any>{
+	withItem(obj: any) {
+		obj.should.equal(divContainer);
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts:32]: Yes, these are all object specific to the tests and not generic at all. Should I just move them to the corresponding test file?

[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts:32]: You might as well make them generic now - they'll be useful to use for other tests so no reason not to do that now so we don't have to come back and do that later on.

File: extensions/arc/src/common/kubeUtils.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	name: string;
+	isCurrentContext: boolean;
+}
+
+export async function getKubeConfigClusterContexts(configFile: string): Promise<KubeClusterContext[]> {
+	try {
+		await fs.promises.access(configFile);
+	} catch (error) {
+		if (error && error.code === 'ENOENT') {
+			return [];
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/common/kubeUtils.ts:21]: I agree, since we are already throwing when it is not access denied, I do not see the need to special case. No sure why @alanrenmsft, special cased it originally in rd.

[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/common/kubeUtils.ts:21]: remove special checking for ENOENT, we will just bubble up any error while accessing the config file.

File: extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts:32]: I though it would be better to pay that cost when we need to use these for something else

[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts:32]: It's not really a cost at this point - it's what should have been done from the start and should be about the same amount of code. 

Part of the benefits of using TypeMoq is you can just fill in the parts you need and then others can easily add in other functionality later on as needed.

File: extensions/arc/src/ui/components/radioOptionsGroup.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+					label: option,
+					checked: option === defaultValue,
+					name: option,
+					value: option,
+					enabled: true
+				}).component();
+				if (radioOption.checked) {
+					this._currentRadioOption = radioOption;
+				}
+				this._onNewDisposableCreated(radioOption.onDidClick(() => {
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/ui/components/radioOptionsGroup.ts:44]: Each radioButton object is inpdependent of other, so I do not know/see what could be un-checking radio button on the previously checked button when a new one is checked. Is there a RadioButtonGroup object now shared through azdata that I should be using instead of individual radio button?

File: extensions/arc/src/test/ui/components/radioOptionsGroup.test.ts (hunk 2 above)
Comment (ranasaria) [reply extensions/arc/src/test/ui/components/radioOptionsGroup.test.ts]: So, I am not sure I am following your comment in the context of the code here. We are just clicking the 2nd radio option within the test radio group to test what out that the radio group's state changes correctly.

File: .vscode/launch.json [hunk 6]
```diff
@@ -227,13 +227,16 @@
...
+				"runtimeExecutable": "${workspaceFolder}/../arct.sh"
 			},
 			"linux": {
 				"runtimeExecutable": "${workspaceFolder}/scripts/test-extensions-unit.sh"
 			},
 			"webRoot": "${workspaceFolder}",
-			"timeout": 45000
+			"timeout": 45000,
+			"env": {
+				"ADS_TEST_GREP": "radio"
```
Comment (Charles-Gagnon) [inline .vscode/launch.json:238]: undo these changes

File: extensions/arc/src/ui/components/radioOptionsGroup.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply extensions/arc/src/ui/components/radioOptionsGroup.ts:44]: https://www.w3schools.com/tags/tryit.asp?filename=tryhtml5_input_type_radio

Note that the state of the radio buttons are done as a unit - this is functionality provided by using the "name" property to group the buttons together. 

But I don't think we currently utilize this currently.

The ideal behavior - the ModelView components understand this and hook into the events to properly update the components with their enabled state as the buttons are selected/unselected. This is what I was referring to with my first comment. 

I seem to recall looking at this before though and finding out that there isn't any event hooked up to when a radio button was unselected so it would be pretty difficult to make sure the state for each button is kept correctly updated. 

So it'd be good to verify that yourself now - and at the very least this kind of logic to update the checked property shouldn't be necessary with how it looks like you have this set up since you don't actually care about keeping the checked property up to date. You just care about whatever the last one selected was.

File: extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts]: The main benefit of TypeMoq is that it allows one to construct a mock object for an Interface which sinon does not allow. Otherwise, sinon is easier to work with and particularly easier to troubleshoot and debug when things do not work as per plan.

[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts]: Changed to using generic typemoq mock objects. Removed all sinon mocks which were specific to the tests being written. A nice result was that the test code has no typemoq/sinon code and the stub where all the typemoq code resides has on test knowledge.

File: extensions/arc/src/ui/components/radioOptionsGroup.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/ui/components/radioOptionsGroup.ts:47]: Ok will check if it works if I start setting the name property to group the buttons together.

File: extensions/arc/src/common/kubeUtils.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+		}
+		else {
+			throw error;
+		}
+	}
+	const config = yamljs.load(configFile);
+	const rawContexts = <any[]>config['contexts'];
+	const currentContext = <string>config['current-context'];
+	const contexts: KubeClusterContext[] = [];
+	if (currentContext && rawContexts && rawContexts.length > 0) {
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/common/kubeUtils.ts]: changed to throw an exception.

File: extensions/arc/src/common/kubeUtils.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+		rawContexts.forEach(rawContext => {
+			const name = <string>rawContext['name'];
+			if (name) {
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/common/kubeUtils.ts]: changed to throw an exception.

File: extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+import * as should from 'should';
+import { FakeRadioButton } from './fakeRadioButton';
+
+export const radioButtons: FakeRadioButton[] = [];
+
+export const divContainer = <azdata.DivContainer><any>{
+	withProperties(obj: azdata.DivContainerProperties) {
+		should(obj).not.be.undefined();
+		should(obj.clickable).not.be.undefined();
+		obj.clickable!.should.equal(false, 'withProperties need to be get called with clickable set to false');
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts]: Removed fakeContainersAndBuilders.ts file. New Typemoq stubs created in stubs.ts to provide equivalent functionality.

File: extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	addItem: (component: azdata.Component) => radioButtons.push(component as FakeRadioButton),
+	get items() { return radioButtons; }
+};
+
+export const loadingComponent = <azdata.LoadingComponent>{
+	get loading() { return false; },
+	set loading(_value: boolean) { }
+};
+
+const loadingBuilder = (<azdata.LoadingComponentBuilder><any>{
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/test/mocks/fakeContainersAndBuilders.ts]: Removed.

File: extensions/arc/src/ui/components/radioOptionsGroup.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+		this._divContainer.clearItems();
+		try {
+			const optionsInfo = await optionsInfoGetter();
+			const options = optionsInfo.values!;
+			let defaultValue: string = optionsInfo.defaultValue!;
+			options.forEach((option: string) => {
+				const radioOption = this._view!.modelBuilder.radioButton().withProperties<azdata.RadioButtonProperties>({
+					label: option,
+					checked: option === defaultValue,
+					name: option,
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/ui/components/radioOptionsGroup.ts]: Setting name to a unique string that corresponds to the RadioOptionsGroup name.

File: extensions/arc/src/ui/components/radioOptionsGroup.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/ui/components/radioOptionsGroup.ts:50]: Ok removed forEach loop to check all the radio buttons. Checked on your suggestion, if we do not set checked to false of previously checked radio Button, the UI and all functions work fine. It is a little weird that a previously checked radio button continues to store 'checked' as true in that scenario though everything works. While storing this._currentRadioOption, I am not simply unchecking the state of the previous radio option.

File: extensions/arc/src/test/ui/components/radioOptionsGroup.test.ts (hunk 1 above)
Comment (ranasaria) [reply extensions/arc/src/test/ui/components/radioOptionsGroup.test.ts]: done

File: extensions/arc/src/test/ui/components/radioOptionsGroup.test.ts (hunk 2 above)
Comment (ranasaria) [reply extensions/arc/src/test/ui/components/radioOptionsGroup.test.ts]: No change here, can you give a rough sample of what you are looking to happen here ...

File: .vscode/launch.json (hunk 6 above)
Comment (ranasaria) [reply .vscode/launch.json]: done

File: extensions/arc/src/ui/components/radioOptionsGroup.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply extensions/arc/src/ui/components/radioOptionsGroup.ts:50]: Sounds good. It would be nice if the state was updated but that should be a thing handled on the main thread side anyways if we do end up handling it I would think so doing it here wasn't the greatest anyways.

File: extensions/arc/src/ui/components/radioOptionsGroup.ts [hunk 12]
```diff
@@ -38,16 +40,17 @@ export class RadioOptionsGroup {
...
 					enabled: true
 				}).component();
 				if (radioOption.checked) {
 					this._currentRadioOption = radioOption;
 				}
 				this._onNewDisposableCreated(radioOption.onDidClick(() => {
-					this._divContainer.items
-						.filter(otherOption => otherOption !== radioOption)
-						.forEach(otherOption => (otherOption as azdata.RadioButtonComponent).checked = false);
+					// this._divContainer.items
```
Comment (Charles-Gagnon) [inline extensions/arc/src/ui/components/radioOptionsGroup.ts]: remove

**Review by Charles-Gagnon — APPROVED** [review]

File: extensions/arc/src/common/utils.ts [hunk 13]
```diff
@@ -282,8 +282,18 @@ export function convertToGibibyteString(value: string): string {
  * @param condition
  * @param message
  */
-export function throwUnless(condition: boolean, message?: string): asserts condition {
+export function throwUnless(condition: any, message?: string): asserts condition {
 	if (!condition) {
```
Comment (alanrenmsft) [inline extensions/arc/src/common/utils.ts:286]: if we do this 0 and '' will also cause it to throw, is that expected or you want to explicitly check undefined:
```
if (condition !== undefined)?
```

File: extensions/arc/src/test/mocks/fakeRadioButton.ts [hunk 14]
```diff
@@ -0,0 +1,91 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as azdata from 'azdata';
+import * as vscode from 'vscode';
+
+export class FakeRadioButton implements azdata.RadioButtonComponent {
```
Comment (alanrenmsft) [inline extensions/arc/src/test/mocks/fakeRadioButton.ts:9]: we usually use 'Mock' instead 'Fake'

File: extensions/arc/src/ui/components/radioOptionsGroup.ts [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+	constructor(private _view: azdata.ModelView, private _onNewDisposableCreated: (disposable: vscode.Disposable) => void, private _groupName: string = `RadioOptionsGroup${RadioOptionsGroup.id++}`) {
+		const divBuilder = this._view.modelBuilder.divContainer();
+		const divBuilderWithProperties = divBuilder.withProperties<azdata.DivContainerProperties>({ clickable: false });
+		this._divContainer = divBuilderWithProperties.component();
+		const loadingComponentBuilder = this._view.modelBuilder.loadingComponent();
+		this._loadingBuilder = loadingComponentBuilder.withItem(this._divContainer);
+	}
+
+	public component(): azdata.LoadingComponent {
```
Comment (alanrenmsft) [inline extensions/arc/src/ui/components/radioOptionsGroup.ts:28]: you meant to make this a getter, right? 

```
public get component(): azdata.LoadingComponent
```

File: extensions/arc/src/ui/components/radioOptionsGroup.ts [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+			this._divContainer.addItem(errorLabel);
+		}
+		this.component().loading = false;
+	}
+
+	get value(): string | undefined {
+		return this._currentRadioOption?.label;
+	}
+
+	get checked(): azdata.RadioButtonComponent {
```
Comment (alanrenmsft) [inline extensions/arc/src/ui/components/radioOptionsGroup.ts:73]: 1. checked should be a boolean, how about selectedOption
2. in Line 70, it is possible that this._currentRadioOption could be undefined, but here I don't see undefined as a possible type
3. is the value getter really needed, user can simply do .selectedOption?.label

File: extensions/arc/src/ui/components/radioOptionsGroup.ts (hunk 15 above)
Comment (ranasaria) [reply extensions/arc/src/ui/components/radioOptionsGroup.ts:28]: Just trying to keep usage consistent with azdata.component methods. It makes for easier usage.

File: extensions/arc/src/test/mocks/fakeRadioButton.ts (hunk 14 above)
Comment (ranasaria) [reply extensions/arc/src/test/mocks/fakeRadioButton.ts:9]: There are already Fake AzdataApi and faceControllerModel objects here, do we need to change them all? Any particular reason to use one name above the other?

**Review by alanrenmsft — APPROVED** [review]

File: extensions/arc/src/common/utils.ts (hunk 13 above)
Comment (ranasaria) [reply extensions/arc/src/common/utils.ts:286]: Yes, that is intended and hence the change. If the caller wants to not do that they can always pass in a more specific condition. The previous implementation was unnecessarily restrictive, the new allows the caller full freedom to pass-in the condition checks as needed.

File: extensions/arc/src/ui/components/radioOptionsGroup.ts (hunk 12 above)
Comment (ranasaria) [reply extensions/arc/src/ui/components/radioOptionsGroup.ts]: You were somehow looking at an old iteration as this was already removed.

File: extensions/arc/src/ui/components/radioOptionsGroup.ts (hunk 16 above)
Comment (ranasaria) [reply extensions/arc/src/ui/components/radioOptionsGroup.ts:73]: Good catch on name and return type. Fixed.

File: extensions/arc/src/ui/components/radioOptionsGroup.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/arc/src/ui/components/radioOptionsGroup.ts:50]: If you think it is better not to do the update here then I can leave just a comment here without the corresponding update. We can do the right fix on the main thread separately, I do not want to pull that change in here.

File: extensions/arc/src/test/stubs.ts [hunk 17]
```diff
@@ -3,8 +3,54 @@
  *  Licensed under the Source EULA. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
+import * as azdata from 'azdata';
+import * as TypeMoq from 'typemoq';
 import * as vscode from 'vscode';
 
+export function createModelViewMock() {
```
Comment (ranasaria) [inline extensions/arc/src/test/stubs.ts:10]: The basic template of these methods to create TypeMock mocks were share by @Charles-Gagnon. These were helpful in factoring the test code right. I am very thankful of his help here.
